gym/vector/async_vector_env.py

```diff
@@ -0,0 +1,482 @@
+        behaviour. If `False`, then the environments call `reset` at the end of
+        each episode.
+
+    context : str, optional
+        Context for multiprocessing. If `None`, then the default context is used.
+        Only available in Python 3.
+    """
+    def __init__(self, env_fns, observation_space=None, action_space=None,
+                 shared_memory=True, copy=True, max_retries=0,
+                 episodic=False, context=None):
+        try:
+            self.ctx = mp.get_context(context)
+        except AttributeError:
+            logger.warn('Context switching for `multiprocessing` is not '
+                'available in Python 2. Using the default context.')
+            self.ctx = mp
+        self.env_fns = env_fns
+        self.shared_memory = shared_memory
+        self.copy = copy
+        self.max_retries = max_retries
+        self.episodic = episodic
+        self._num_retries = 0
+
+        if (observation_space is None) or (action_space is None):
+            dummy_env = env_fns[0]()
+            observation_space = observation_space or dummy_env.observation_space
+            action_space = action_space or dummy_env.action_space
+            dummy_env.close()
+            del dummy_env
+        super(AsyncVectorEnv, self).__init__(num_envs=len(env_fns),
+            observation_space=observation_space, action_space=action_space)
+
+        if self.shared_memory:
+            self._obs_buffer = create_shared_memory(
+                self.single_observation_space, n=self.num_envs)
+            self.observations = read_from_shared_memory(self._obs_buffer,
+                self.single_observation_space, n=self.num_envs)
+        else:
+            self._obs_buffer = None
+            self.observations = create_empty_array(
+                self.single_observation_space, n=self.num_envs, fn=np.empty)
+
+        self.parent_pipes, self.processes = [], []
+        self.error_queue = self.ctx.Queue()
+        with clear_mpi_env_vars():
+            for index in range(self.num_envs):
+                process, parent_pipe = self._start_process(index)
+                self.parent_pipes.append(parent_pipe)
+                self.processes.append(process)
+
+        self.waiting_reset = False
+        self.waiting_step = False
+        self._check_observation_spaces()
+
+    def seed(self, seeds=None):
+        self._assert_is_running()
+        if seeds is None:
+            seeds = [None for _ in range(self.num_envs)]
+        if isinstance(seeds, int):
+            seeds = [seeds + i for i in range(self.num_envs)]
+        assert len(seeds) == self.num_envs
+
+        for pipe, seed in zip(self.parent_pipes, seeds):
+            pipe.send(('seed', seed))
+        for pipe in self.parent_pipes:
+            pipe.recv()
+
+    def reset_async(self):
+        self._assert_is_running()
+        if self.waiting_step:
+            logger.warn('Calling `reset_async` while waiting for a pending '
```

Comment:
Is there any reason to not make this a fatal error?

Reply:
This also applies to all the other cases where it looks like the user called the methods out of order.  Rather than trying to fix it, this should just throw an exception so the user can fix their code.

Reply:
No reason not to make it fatal, I will throw an exception